### DISCUSSION TITLE
How to locally clone and test pull requests

### DISCUSSION THREAD

**Discussion by peepo5** [FreeTubeApp/FreeTube#1734] [General]
### Dependencies
Requires `git` and `gh`.
If you don't have any experience with the terminal, on windows there is `microsoft-windows-terminal` and on linux, just open any terminal.
On windows, installing git and gh requires `chocolatey` or a similar package manager.
### Cloning FreeTube
If you do not have FreeTube locally yet, clone it with this:
```sh
gh repo clone FreeTubeApp/FreeTube
cd FreeTube
```
Install dependencies:
```sh
yarn install
```
### Get the PR you want

Checkout the PR (clones and switches to the branch)

```sh
gh pr checkout [pr number]
```

It will automatically switch to the cloned PR. If you want to change the branch again (example):

```sh
git checkout master
```

**Comment by efb4f5ff-1298-471a-8973-3d47447115dc** [comment]
Lgtm, didnt try it yet but looks good. @peepopoggers for the win!

I think its great to include this in the documentation somewhere for future testers/devs

**Comment by ChunkyProgrammer** [comment]
Isn't yarn being used to install now?

**Reply by peepo5** [reply]
O shoot :smile: 
Updated

**Reply by efb4f5ff-1298-471a-8973-3d47447115dc** [reply]
so i was trying to follow this exactly but when i wanted to switch back to master i could't

https://stackoverflow.com/questions/15745045/how-do-i-resolve-git-saying-commit-your-changes-or-stash-them-before-you-can-me/15745424#15745424

i think this needs to be added.

**Reply by peepo5** [reply]
did you download master? I thought development was the default branch

**Reply by efb4f5ff-1298-471a-8973-3d47447115dc** [reply]
i did. the second and the third time i followed ur steps this didnt happen. I think that i made a change by accident that triggered this error. I think its worth putting in

**Comment by efb4f5ff-1298-471a-8973-3d47447115dc** [comment]
Ah i saw u updated the guide.

I tried it out and after i typed the clone command it prompted me to do

```
gh auth login
```

U have to do a few steps here in order to actually start cloning and i in my opinion a bit overwhelming for somebody new trying to test a PR.

**What is ur preferred protocol for Git operation**
- HTTPS
- SSH

What if u dont know? The guide should tell u what option to take if u dont know.

If u choose what protocol u preferred u get the question

**Authenticate Git with ur Github credentials**
- Y/n

I tried n and it still prompted me to login.

I think we need really need to guide new people in this process otherwise they might think its too difficult to pull a PR and test it. Im really trying to do everything like a noob so this will be fool proof.

**Reply by efb4f5ff-1298-471a-8973-3d47447115dc** [reply]
So i chose HTTPS and authenticated myself and tried cloning again. Ended up with this error. UX is falling through the roof :(

`Cloning into 'FreeTube'...
warning: templates not found /usr/share/git-core/templates
fatal: Unable to find remote helper for 'https'
exit status 128
`

**Reply by efb4f5ff-1298-471a-8973-3d47447115dc** [reply]
The method before u updated this was way better in my opinion

**Reply by peepo5** [reply]
Yes but then that method then made the prs hard to update

**Reply by efb4f5ff-1298-471a-8973-3d47447115dc** [reply]
How did it made the PR hard to update?

**Reply by peepo5** [reply]
git pull origin <prname> did not work

**Reply by efb4f5ff-1298-471a-8973-3d47447115dc** [reply]
Thats weird because i just used it to pull Chunkys PR about comments

**Comment by absidue** [comment]
Adding a note here for anyone else trying to build FreeTube on node 17, it doesn't work at the moment so the only solution, apart from updating the dependencies, is to downgrade to the latest version of node 16 (at the time of writing 16.13.0).

**The problem:**
Node 17 switched to using OpenSSL 3, which has disabled MD4, webpack (and i think vue) were using that, webpack has been updated to fix the problem. The recommended workaround on the webpack issue is to run `export NODE_OPTIONS=--openssl-legacy-provider` before running webpack, that only fixes the build though, FreeTube itself doesn't start with `yarn run dev`.

**Comment by efb4f5ff-1298-471a-8973-3d47447115dc** [comment]
I cant test PR's anymore and i dont know why. Are these steps up to date because they dont seem to work for me.

**Reply by absidue** [reply]
Make sure you are using nodejs 16, 17 doesn't work.

**Reply by efb4f5ff-1298-471a-8973-3d47447115dc** [reply]
Not sure if thats the issue here when i try to fetch PR #1800

It says: unable to access github.com/Freetubeapp/Freetube.git/ could not resolve host github.com

**Reply by absidue** [reply]
Check that you aren't using git protocol URLs `git://` GitHub is phasing those out and slowly disabling access to GitHub through them as they are unencrypted.

**Reply by efb4f5ff-1298-471a-8973-3d47447115dc** [reply]
Before i wanted to do as u suggested i restarted my machine and it now works again.

Thanks for the help and ur time though appreciate it!

**Comment by efb4f5ff-1298-471a-8973-3d47447115dc** [comment]
This always works for me

Clone FreeTube:

```
git clone https://github.com/FreeTubeApp/FreeTube.git
```
Go to the folder:
```
cd FreeTube
```

Add a pull request as a branch (replace square brackets):
```
git fetch origin pull/[pr number]/head:[your name for the pr]
```

Change to the added branch:
```
git checkout [your name for the pr]
```

Install dependencies:
```
yarn
```

Test the PR:
```
yarn run dev
```


Note from Preston: 
>"I would wait until the new branch is checked out before installing dependencies. There's always a chance of something getting changed / removed that could break the app that you wouldn't see otherwise.
